src/grep.js

```diff
@@ -68,6 +68,10 @@ function _grep(options, regex, files) {
     }
   });
 
+  if (grep.length === 0) {
```

Comment:
Ah, so this is an actual bug. There's a few reasons the `grep` array may be empty:

 1. No files (but we exit early for this
 2. Files are specified, but we couldn't read any of them (perhaps files don't exist)
 3. We can read some files but not others. Of the files we can read, none match the regex
 4. We can read all files, but none contain the regex

Case 1 is [already handled](https://github.com/shelljs/shelljs/blob/4bd22e77423182219cd43f0c8d38621b1c957986/src/grep.js#L37). You filed a ticket about case 4.

For cases 2 and 3, the correct behavior is a non-zero exit code, and stderr should say something like `grep: no such file or directory: nameoffile.txt` (repeated for each problematic file). Error code should be 2, as this indicates a problematic file.

---

I think you can change this if-condition to also check `common.state.error` or `common.state.errorCode`. Don't modify this value.

Reply:
Ok, thanks. That’s what I thought was happening. Wasn’t sure what to add to the conditional, that info is very helpful